src/mbgl/map/map.cpp

```diff
@@ -881,18 +882,20 @@ void Map::addLayer(std::unique_ptr<Layer> layer, const optional<std::string>& be
     impl->backend.deactivate();
 }
 
-void Map::removeLayer(const std::string& id) {
+std::unique_ptr<Layer> Map::removeLayer(const std::string& id) {
     if (!impl->style) {
-        return;
+        return nullptr;
     }
 
     impl->styleMutated = true;
     impl->backend.activate();
 
-    impl->style->removeLayer(id);
+    auto removedLayer = impl->style->removeLayer(id);
     impl->onUpdate(Update::Classes);
 
     impl->backend.deactivate();
```

Comment:
We will need to add special logic for `CustomLayer`.  Currently, `~CustomLayer` calls the `CustomLayerDeinitializeFunction`, which may make OpenGL calls. This is the reason that we call `impl->backend.activate()/deactivate()` -- the context needs to be active when initialization/deinitialization happens.

I think we need to preserve the behavior that removing a CustomLayer triggers deinitialization, even when returning ownership. Let's do this in `Style::removeLayer`, making it parallel `Style::addLayer`, which does initialization.